### ISSUE TITLE
Add a Replicate-like HTTP API to Cog

### ISSUE THREAD

**Issue by nickstenning** [replicate/cog#821]
The Cog HTTP API currently has a single prediction-related endpoint:

```
POST /predictions
{
    "input": ...
}
⇒
{
    "output": ...
    "status": "succeeded",
}
```

The call is blocking: it will wait for the prediction to finish before returning a response.

In order to make it easier to coordinate the actions of the Cog container from another process, we want to add an API which has the following properties:

- It returns immediately (i.e. before the prediction has finished processing), giving control back to a hypothetically single-threaded caller.
- It accepts something more like a full prediction object, complete with support for a `webhook` property, so that it can give progress updates to a URL provided by the caller.
- It has a cancelation endpoint which triggers cancelation of the in-flight prediction.

We've made a start on this at https://github.com/replicate/cog/compare/add-async-http-api, which may be helpful as a reference when reading the below.

At the moment we're proposing new endpoints under the `/v1` URL prefix in order to minimize the chance of backwards-incompatible changes, but we might be able to implement these changes on the `/predictions` endpoint without breaking anything if we are happy with some out-of-band signal (e.g. cog startup config, a URI query parameter, or a header) which indicates that we want non-blocking responses.

What we currently propose is as follows:

```
POST /v1/predictions               (only works if no predictions are currently running, 409s otherwise)
GET  /v1/predictions/{id}          (only works for the most recent prediction, 404s otherwise) 
POST /v1/predictions/{id}/cancel   (only works for the most recent prediction, 404s otherwise)
```

Currently the idea is that file outputs would be handled by making the prediction files available over the HTTP API as well, and it would be for the sibling container to handle uploads to GCS/etc.

Questions for others:

1. What do we think of the idea of having prediction IDs in URLs that only work for the most recent prediction?
2. If we didn't do that, what would a cancellation API look like?
3. Do we want/need a GET endpoint, or are webhooks enough?
4. Should we try and build this on top of `/predictions` in a backwards-compatible way? Any thoughts on what the out-of-band signal that we want non-blocking responses should be?

**Comment by nickstenning** [comment]
One additional complication here, actually, is that it seems that the HTTP API is actually designed to be multithreaded for CPU models. That means that "most recent prediction" isn't a well-defined concept, and so we'd presumably have to do something like keep prediction state around for some (configurable?) amount of time before cleaning them up.

**Comment by bfirsh** [comment]
> What do we think of the idea of having prediction IDs in URLs that only work for the most recent prediction?

Weird, innit. Could we not just keep them in memory? With a max number stored, or something. In production, the max number stored could be set to one.

Perhaps we need to think about some use-cases to design how the API should behave. How you might develop with them, how you might deploy them, etc. I'll let this percolate.

> Do we want/need a GET endpoint, or are webhooks enough?

I'd imagine we'd want to mirror the Replicate API, but perhaps we could leave it out for now if we want to ship a minimal thing?

Again, this comes back to use-cases. If you're using it in development, you don't want to set up webhooks.

> Should we try and build this on top of /predictions in a backwards-compatible way? Any thoughts on what the out-of-band signal that we want non-blocking responses should be?

Blocking responses are a convenient way to develop, and I imagine we might want them in Replicate at some point too. I wonder what the API would look like. `?wait=1`? Or the inverse?

**Comment by bfirsh** [comment]
@nickstenning mentioned: Are there two schemas here or one? A single "prediction" or a request and reponse?

**Comment by bfirsh** [comment]
"version" is in the Replicate API. Is there a way we can make that a supported part of the API? Maybe it's in the URL? Maybe it's an "extension" to the API?

**Comment by nickstenning** [comment]
> > Blocking responses are a convenient way to develop, and I imagine we might want them in Replicate at some point too. I wonder what the API would look like. ?wait=1? Or the inverse?
> 
> HTTP has a `Prefer` header that clients can use to request responses to be returned asynchronously. From [RFC 7240](https://www.rfc-editor.org/rfc/rfc7240):


Don't want this get lost so thank you for this @mattt! This is really helpful and there's a good chance it's what we want to do here.

**Comment by nickstenning** [comment]
I'm delighted to say this is closed by #870. There's still plenty more to do here but the difficult stuff is done.
